flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -116,14 +125,60 @@ public void registerScalarFunction(String name, ScalarFunction function) {
 	}
 
 	public String[] getUserDefinedFunctions() {
-		return userFunctions.values().stream()
-			.map(FunctionDefinition::toString)
-			.toArray(String[]::new);
+		List<String> result = new ArrayList<>();
+
+		// Get functions in catalog
+		Catalog catalog = catalogManager.getCatalog(catalogManager.getCurrentCatalog()).get();
+		try {
+			result.addAll(catalog.listFunctions(catalogManager.getCurrentDatabase()));
+		} catch (DatabaseNotExistException e) {
+			// Ignore since there will always be a current database of the current catalog
+		}
+
+		// Get functions registered in memory
+		result.addAll(
+			userFunctions.values().stream()
+				.map(FunctionDefinition::toString)
+				.collect(Collectors.toList()));
+
+		// Get built-in functions
+		result.addAll(
+			BuiltInFunctionDefinitions.getDefinitions()
+				.stream()
+				.map(f -> f.getName())
+				.collect(Collectors.toList()));
+
+		return result.stream()
+			.map(n -> normalizeName(n))
+			.collect(Collectors.toList())
+			.toArray(new String[0]);
 	}
 
 	@Override
 	public Optional<FunctionLookup.Result> lookupFunction(String name) {
```

Comment:
I'm curious about the format of name here. In complete form, a lookup is for a function like catalog.database.functionName, though it's possible to have database.functionName or just functionName. When it's just functionName, it can be either refer to one in current catalog/db or a built-in function.

Reply:
I have the same question here. I guess the signature needs to be changed to `String[] path`. And we cannot simply assume the current catalog but must consider all catalogs available.

Reply:
@twalthr Agree. As I mentioned in the comment, I hope to change this API in a followup PR rather than this PR. As such a change would require refactoring upstream code which will complicate this PR itself.